### PR TITLE
feat(plugin-server): simplify groups caching

### PR DESCRIPTION
**Pull request by yakkomajuri**
I started looking at the groups cache implementation to hunt bugs (see #12371 #12357) and I just realized that the implementation we currently have is so convoluted it's very hard to follow and debug.

It consists of 4 functions, iterates over all groups 5 times (with map, filter, etc), and constantly changes the format of the data. 

I really struggled to grasp what was going on and it took me hours to parse through it all thoroughly. As such, I spent less than an hour writing the current implementation, where everything we need to understand is nicely packed into one function with a better logical structure and comments.

The one "bigger" change I've also made is that I've opted for _n_ (where `n > 0` and `n <= 5`) Postgres queries over 1  single query to fetch all groups like we had before. The reason for this is that it makes things much cleaner and easy to maintain as we already have a `fetchGroup` function. The batch query made things more complicated and required some SQL building that actually initially had a bug (see #12371). 

Also, at best this was an early optimization that we can bring back in the future. If we look at the data, the average number of distinct groups per team is 1.51, and currently we [run the batch query ~24 times per minute](https://app.pganalyze.com/databases/-675880137/queries/5633379409). Thus, the impact of these individual queries for when we miss the cache should be minimal, at least for now. This is something we can definitely reassess.

I've also added some missing metrics.

**Edit**

Ah, and another change is that we won't cache "missing" data as we used to. We can again reassess this decision but as per the metrics I've shared above this should be ok. The problem with caching missing data is that it is very likely to cause some events that should have groups data to not have it. If a group doesn't exist ("missing data") yet we're getting events for it, it means it is _about to_ exist, and thus caching the missing data would a) not be very impactful and b) cause us to not add data to certain events. 
 
Overall, let's start simple with something we can understand and optimize as necessary.

### CODE REVIEW

File: plugin-server/src/utils/db/db.ts [hunk 1]
```diff
@@ -673,109 +654,70 @@ export class DB {
...
-
-        if (data === 'missing' || !data) {
-            return { identifier: groupIdentifier, data: null, cached: !data }
-        }
-        return {
-            identifier: groupIdentifier,
-            data: { properties: data.properties, created_at: DateTime.fromISO(data.created_at).toUTC() },
-            cached: true,
-        }
+    public async updateGroupCache(
```
Comment (yakkomajuri) [inline plugin-server/src/utils/db/db.ts:657]: Here's a cleaner look at the new function:

https://github.com/PostHog/posthog/blob/71649cbea619b1de13cb2e3fa993e73c26228b23/plugin-server/src/utils/db/db.ts#L657

**Comment by yakkomajuri** [comment]
This is reviewable, although I'm still debugging some tests

**Comment by yakkomajuri** [comment]
Damn, thought I'd tagged you both for review earlier @hazzadous @tiina303. Have some test failures that I'm dealing with but the core of this should be reviewable

**Review by hazzadous — APPROVED** [review]
All for simpler 💯 

... with passing tests of course 👍

**Comment by tiina303** [comment]
> Ah, and another change is that we won't cache "missing" data as we used to. We can again reassess this decision but as per the metrics I've shared above this should be ok. The problem with caching missing data is that it is very likely to cause some events that should have groups data to not have it. If a group doesn't exist ("missing data") yet we're getting events for it, it means it is about to exist, and thus caching the missing data would a) not be very impactful and b) cause us to not add data to certain events.

We don't necessarily have that group info we only create a group and info on `$groupidentify` events
https://github.com/PostHog/posthog/blob/master/plugin-server/src/worker/ingestion/process-event.ts#L145-L147

So we'll potentially be hammering Postgres if we don't cache that we're missing data.

+1 for simpler is better.

File: plugin-server/src/utils/db/db.ts [hunk 2]
```diff
@@ -673,109 +654,70 @@ export class DB {
...
+                const cachedGroupData = await this.redisGet<CachedGroupData | null>(groupCacheKey, null)
+
+                if (cachedGroupData) {
+                    this.statsd?.increment('group_info_cache.hit')
+                    groupColumns[propertiesColumnName] = JSON.stringify(cachedGroupData.properties)
+                    groupColumns[createdAtColumnName] = cachedGroupData.created_at
+                    continue
+                }
+            } catch (error) {
+                captureException(error)
```
Comment (tiina303) [inline plugin-server/src/utils/db/db.ts:686]: maybe add a debug log too?

File: plugin-server/src/utils/db/db.ts [hunk 3]
```diff
@@ -673,109 +654,70 @@ export class DB {
...
+            this.statsd?.increment('group_info_cache.miss')
 
-    public async fetchGroupColumnsValues(teamId: number, groups: GroupIdentifier[]): Promise<Record<string, string>> {
-        if (groups.length === 0) {
-            return {}
-        }
-        const cachedResults = await Promise.all(
-            groups.map((groupIdentifier) => this.getGroupDataCache(teamId, groupIdentifier))
-        )
+            // If we didn't find cached data, lookup the group from Postgres
```
Comment (tiina303) [inline plugin-server/src/utils/db/db.ts:691]: nit: instead of commenting consider having separate functions for getting data from cache and lookup from Postgres, if the function is short and calls clearly named functions the comments aren't needed

File: plugin-server/src/utils/db/db.ts [hunk 4]
```diff
@@ -673,109 +654,70 @@ export class DB {
...
+                try {
+                    await this.updateGroupCache(teamId, groupTypeIndex, groupKey, {
+                        properties: storedGroupData.group_properties,
+                        created_at: createdAt,
+                    })
+                } catch (error) {
+                    captureException(error)
+                }
+            } else {
+                // We couldn't find the data from the cache nor Postgres, so record this in a metric and in Sentry
```
Comment (tiina303) [inline plugin-server/src/utils/db/db.ts:711]: this can happen because we only create the group if `groupidentify` is called ... this makes me think should we change that and always create a group if it doesn't exist? And furthermore if we did that we could pass the group info forward similar to person info and not need to have caching here at all either?

File: plugin-server/src/utils/db/db.ts [hunk 5]
```diff
@@ -673,109 +654,70 @@ export class DB {
...
+                this.statsd?.increment('groups_data_missing_entirely')
+                captureException(new Error('Missing groups data entirely'), { extra: { groupCacheKey } })
+
+                groupColumns[propertiesColumnName] = '{}'
```
Comment (tiina303) [inline plugin-server/src/utils/db/db.ts:715]: what if we added a property that's internal telling us we hit this, so it will be easier for us to distinguish these cases from other empty ones?

**Review by tiina303 — APPROVED** [review]

**Comment by yakkomajuri** [comment]
> We don't necessarily have that group info we only create a group and info on $groupidentify events

Right, but `$groupidentify` is what attaches the group properties to the event in the first place. Sure, one could send it manually, but in that case we should assume: 

a) They called `$groupidentify` somewhere else e.g. client and sent the event from the backend
b) Have a wrong implementation - I guess for this we could have an ingestion warning. 

I'll monitor when I ship and consider how we could do this without being subject to race conditions

File: plugin-server/src/utils/db/db.ts (hunk 5 above)
Comment (yakkomajuri) [reply plugin-server/src/utils/db/db.ts:715]: I've thought about these things before, but feels... hacky

File: plugin-server/src/utils/db/db.ts (hunk 4 above)
Comment (yakkomajuri) [reply plugin-server/src/utils/db/db.ts:711]: Yeah I think always creating like we do with persons is a good idea. Worth trying to find out why we didn't do this in the first place

File: plugin-server/src/utils/db/db.ts (hunk 3 above)
Comment (yakkomajuri) [reply plugin-server/src/utils/db/db.ts:691]: I'm generally a proponent of this but in this case this function is small enough and "integrated" enough (setting/updating state) that I think this is better than having to jump all around the place to understand what's going on

**Comment by yakkomajuri** [comment]
merging and monitoring

File: plugin-server/src/utils/db/db.ts (hunk 4 above)
Comment (tiina303) [reply plugin-server/src/utils/db/db.ts:716]: persons was changed recently with the changes to the pipeline, I think this is just historical.
